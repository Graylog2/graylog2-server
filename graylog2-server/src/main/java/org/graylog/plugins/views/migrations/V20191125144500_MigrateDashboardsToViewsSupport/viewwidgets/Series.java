/**
 * This file is part of Graylog.
 *
 * Graylog is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Graylog is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Graylog.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.graylog.plugins.views.migrations.V20191125144500_MigrateDashboardsToViewsSupport.viewwidgets;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.google.auto.value.AutoValue;
import org.graylog.plugins.views.migrations.V20191125144500_MigrateDashboardsToViewsSupport.SeriesSpec;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

@AutoValue
public abstract class Series {
    private static final String FIELD_CONFIG = "config";
    private static final String FIELD_FUNCTION = "function";

    private static final Pattern destructuringPattern = Pattern.compile("(\\w+)\\((\\w+)?\\)");

    @JsonProperty(FIELD_CONFIG)
    public abstract SeriesConfig config();

    @JsonProperty(FIELD_FUNCTION)
    public abstract String function();

    public static Builder builder() {
        return new AutoValue_Series.Builder()
                .config(SeriesConfig.empty());
    }

    public static Builder buildFromString(String function) {
        return builder().function(function).config(SeriesConfig.empty());
    }

    public static Series create(String function, String field) {
        return buildFromString(function + "(" + field + ")").build();
    }

    public SeriesSpec toSeriesSpec() {
        final Matcher matcher = destructuringPattern.matcher(function());
        if (matcher.matches()) {
            final String functionName = matcher.group(1);
            final String optionalFieldName = matcher.group(2);
            return SeriesSpec.create(functionName, function(), optionalFieldName);
        }

        throw new RuntimeException("Unable to parse function: " + function());
    }

    @AutoValue.Builder
    public static abstract class Builder {
        public abstract Builder config(SeriesConfig config);
        public abstract Builder function(String function);
        public abstract Series build();

    }
}
