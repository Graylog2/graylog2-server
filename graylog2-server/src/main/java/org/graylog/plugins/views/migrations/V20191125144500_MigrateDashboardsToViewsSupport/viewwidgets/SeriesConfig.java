/**
 * This file is part of Graylog.
 *
 * Graylog is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Graylog is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Graylog.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.graylog.plugins.views.migrations.V20191125144500_MigrateDashboardsToViewsSupport.viewwidgets;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.google.auto.value.AutoValue;

import javax.annotation.Nullable;

@AutoValue
public abstract class SeriesConfig {
    static final String FIELD_NAME = "name";

    public static SeriesConfig empty() {
        return builder().build();
    }

    @JsonProperty(FIELD_NAME)
    @Nullable
    public abstract String name();

    public static Builder builder() {
        return new AutoValue_SeriesConfig.Builder();
    }

    @AutoValue.Builder
    public static abstract class Builder {
        public abstract Builder name(String name);

        public abstract SeriesConfig build();
    }
}
