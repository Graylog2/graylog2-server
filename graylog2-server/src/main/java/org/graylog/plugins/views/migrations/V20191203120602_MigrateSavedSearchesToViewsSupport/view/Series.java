/**
 * This file is part of Graylog.
 *
 * Graylog is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Graylog is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Graylog.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.graylog.plugins.views.migrations.V20191203120602_MigrateSavedSearchesToViewsSupport.view;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.google.auto.value.AutoValue;
import org.graylog.plugins.views.migrations.V20191203120602_MigrateSavedSearchesToViewsSupport.search.SeriesSpec;

import java.util.Collections;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@AutoValue
public abstract class Series {
    private static final String FIELD_CONFIG = "config";
    private static final String FIELD_FUNCTION = "function";

    private static final Pattern destructuringPattern = Pattern.compile("(\\w+)\\((\\w+)?\\)");

    @JsonProperty(FIELD_CONFIG)
    public Map<String, Object> config() {
        return Collections.emptyMap();
    }

    @JsonProperty(FIELD_FUNCTION)
    public String function() {
        return "count()";
    }

    public static Series create() {
        return new AutoValue_Series();
    }

    public SeriesSpec toSeriesSpec() {
        final Matcher matcher = destructuringPattern.matcher(function());
        if (matcher.matches()) {
            final String functionName = matcher.group(1);
            final String optionalFieldName = matcher.group(2);
            return SeriesSpec.create(functionName, function(), optionalFieldName);
        }

        throw new RuntimeException("Unable to parse function: " + function());
    }
}
