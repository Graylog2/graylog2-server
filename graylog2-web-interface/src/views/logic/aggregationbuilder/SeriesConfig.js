// @flow strict
import * as Immutable from 'immutable';

export type SeriesConfigJson = {
  name: string,
};

type InternalState = {
  name: ?string,
};

export default class SeriesConfig {
  _value: InternalState;

  constructor(name: ?string) {
    this._value = { name };
  }

  get name() {
    return this._value.name;
  }

  toJSON() {
    const { name } = this._value;

    return { name };
  }

  static fromJSON(value: SeriesConfigJson) {
    const { name } = value;

    return new SeriesConfig(name);
  }

  static empty() {
    return new SeriesConfig(null);
  }

  toBuilder() {
    // eslint-disable-next-line no-use-before-define
    return new Builder(Immutable.Map(this._value));
  }
}

type BuilderState = Immutable.Map<string, any>;
class Builder {
  value: BuilderState;

  constructor(value: BuilderState = Immutable.Map()) {
    this.value = value;
  }

  name(newName: string) {
    return new Builder(this.value.set('name', newName));
  }

  build() {
    const { name } = this.value.toObject();

    return new SeriesConfig(name);
  }
}
