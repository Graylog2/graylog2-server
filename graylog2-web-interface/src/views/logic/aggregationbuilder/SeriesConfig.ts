/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import * as Immutable from 'immutable';

export type SeriesConfigJson = {
  name: string,
};

type InternalState = {
  name: string | undefined | null,
};

export default class SeriesConfig {
  private readonly _value: InternalState;

  constructor(name: string | undefined | null) {
    this._value = { name };
  }

  get name() {
    return this._value.name;
  }

  toJSON() {
    const { name } = this._value;

    return { name };
  }

  static fromJSON(value: SeriesConfigJson) {
    const { name } = value;

    return new SeriesConfig(name);
  }

  static empty() {
    return new SeriesConfig(null);
  }

  toBuilder() {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    return new Builder(Immutable.Map(this._value));
  }
}

type BuilderState = Immutable.Map<string, any>;

class Builder {
  private readonly value: BuilderState;

  constructor(value: BuilderState = Immutable.Map()) {
    this.value = value;
  }

  name(newName: string) {
    return new Builder(this.value.set('name', newName));
  }

  build() {
    const { name } = this.value.toObject();

    return new SeriesConfig(name);
  }
}
