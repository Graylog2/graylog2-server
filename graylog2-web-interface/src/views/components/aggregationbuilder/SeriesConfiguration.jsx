import React from 'react';
import PropTypes from 'prop-types';
import { trim } from 'lodash';

import { Button, ControlLabel, FormControl, FormGroup, HelpBlock } from 'components/graylog';
import Series from 'views/logic/aggregationbuilder/Series';

export default class SeriesConfiguration extends React.Component {
  static propTypes = {
    series: PropTypes.instanceOf(Series).isRequired,
    onClose: PropTypes.func.isRequired,
    usedNames: PropTypes.arrayOf(PropTypes.string),
  };

  static defaultProps = {
    usedNames: [],
  };

  constructor(props, context) {
    super(props, context);

    this.state = {
      series: props.series,
      name: props.series.effectiveName,
    };
  }

  _onSubmit = () => {
    const { name, series, series: { config, function: functionName, effectiveName } } = this.state;
    const { onClose } = this.props;
    const newName = name || functionName;

    if (newName && newName !== effectiveName) {
      const newConfig = config.toBuilder().name(newName).build();
      const newSeries = series.toBuilder().config(newConfig).build();
      onClose(newSeries);
    } else {
      onClose(series);
    }
  };

  _changeName = e => this.setState({ name: e.target.value });

  render() {
    const { name } = this.state;
    const { usedNames = [] } = this.props;
    const isValid = !usedNames.includes(trim(name));
    const validationHint = isValid ? null : <React.Fragment> <strong>Name must be unique.</strong></React.Fragment>;
    return (
      <span>
        <FormGroup validationState={isValid ? null : 'error'}>
          <ControlLabel>Name</ControlLabel>
          <FormControl type="text" value={name} onChange={this._changeName} />
          <HelpBlock>The name of the series as it appears in the chart.{validationHint}</HelpBlock>
        </FormGroup>
        <div className="pull-right" style={{ marginBottom: '10px' }}>
          <Button bsStyle="success" disabled={!isValid} onClick={this._onSubmit}>Done</Button>
        </div>
      </span>
    );
  }
}
