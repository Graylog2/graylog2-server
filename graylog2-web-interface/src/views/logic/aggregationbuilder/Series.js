// @flow strict
import { get } from 'lodash';
import * as Immutable from 'immutable';

import SeriesConfig from './SeriesConfig';
import type { SeriesConfigJson } from './SeriesConfig';

export type SeriesJson = {
  config: SeriesConfigJson,
  function: string,
};

type InternalState = {
  config: SeriesConfig,
  function: string,
};

export type Definition = {
  type: string,
  field?: string,
  percentile?: string,
};

const parametersRegex = /\((.+)\)/;
const funcNameRegex = /(\w+)\(/;
const testSeriesRegex = /^(\w+)\((\w*)(,(\w+))*\)$/;

const definitionFor = (type: string, parameters: Array<string>): Definition => {
  // eslint-disable-next-line no-unused-vars
  const [field, parameter] = parameters;
  if (type === 'percentile') {
    return { type, field, percentile: parameter };
  }
  return { type, field };
};

export const isFunction = (s: string) => testSeriesRegex.test(s);
export const parseSeries = (s: string) => {
  const funcNameResult = funcNameRegex.exec(s);
  if (!funcNameResult) {
    return null;
  }
  const type = funcNameResult[1];
  const definition: Definition = {
    type,
  };

  const parameterResult = parametersRegex.exec(s);
  if (!parameterResult) {
    return definition;
  }

  return definitionFor(type, parameterResult[1].split(','));
};

export default class Series {
  _value: InternalState;

  constructor(func: string, config: SeriesConfig = SeriesConfig.empty()) {
    this._value = { function: func, config };
  }

  get function() {
    return this._value.function;
  }

  get config() {
    return this._value.config;
  }

  get effectiveName(): string {
    const overridenName = get(this, 'config.name');
    return overridenName || this.function;
  }

  toString() {
    return `Series: ${this.effectiveName}, config={${JSON.stringify(this.config)}}`;
  }

  toJSON() {
    return {
      config: this._value.config,
      function: this._value.function,
    };
  }

  static fromJSON(value: SeriesJson) {
    return new Series(value.function, SeriesConfig.fromJSON(value.config));
  }

  static forFunction(func: string) {
    // eslint-disable-next-line no-use-before-define
    return new Builder()
      .function(func)
      .config(SeriesConfig.empty())
      .build();
  }

  toBuilder() {
    // eslint-disable-next-line no-use-before-define
    return new Builder(Immutable.Map(this._value));
  }
}

type BuilderState = Immutable.Map<string, any>;
class Builder {
  value: BuilderState;

  constructor(value: BuilderState = Immutable.Map()) {
    this.value = value;
  }

  function(newFunction: string) {
    return new Builder(this.value.set('function', newFunction));
  }

  config(newConfig: SeriesConfig) {
    return new Builder(this.value.set('config', newConfig));
  }

  parameter(newParameter: *) {
    return new Builder(this.value.set('parameter', newParameter));
  }

  build() {
    const { config } = this.value.toObject();
    const func = this.value.get('function');
    return new Series(func, config);
  }
}
