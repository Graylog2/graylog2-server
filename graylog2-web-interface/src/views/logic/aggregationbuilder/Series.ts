/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import { get } from 'lodash';
import * as Immutable from 'immutable';

import SeriesConfig from './SeriesConfig';
import type { SeriesConfigJson } from './SeriesConfig';

export type SeriesJson = {
  config: SeriesConfigJson,
  function: string,
};

type InternalState = {
  config: SeriesConfig,
  function: string,
};

export type Definition = {
  type: string,
  field?: string,
  percentile?: string,
};

const parametersRegex = /\((.+)\)/;
const funcNameRegex = /(\w+)\(/;
const testSeriesRegex = /^(\w+)\((.*)(,(\w+))*\)$/;

const definitionFor = (type: string, parameters: Array<string>): Definition => {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const [field, parameter] = parameters;

  if (type === 'percentile') {
    return { type, field, percentile: parameter };
  }

  return { type, field };
};

export const isFunction = (s: string) => testSeriesRegex.test(s);

export const parseSeries = (s: string) => {
  const funcNameResult = funcNameRegex.exec(s);

  if (!funcNameResult) {
    return null;
  }

  const type = funcNameResult[1];
  const definition: Definition = {
    type,
  };

  const parameterResult = parametersRegex.exec(s);

  if (!parameterResult) {
    return definition;
  }

  return definitionFor(type, parameterResult[1].split(','));
};

export default class Series {
  private readonly _value: InternalState;

  constructor(func: string, config: SeriesConfig = SeriesConfig.empty()) {
    this._value = { function: func, config };
  }

  get function() {
    return this._value.function;
  }

  get config() {
    return this._value.config;
  }

  get effectiveName(): string {
    const overridenName = get(this, 'config.name');

    return overridenName || this.function;
  }

  toString() {
    return `Series: ${this.effectiveName}, config={${JSON.stringify(this.config)}}`;
  }

  toJSON() {
    return {
      config: this._value.config,
      function: this._value.function,
    };
  }

  static fromJSON(value: SeriesJson) {
    return new Series(value.function, SeriesConfig.fromJSON(value.config));
  }

  static forFunction(func: string) {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    return new Builder()
      .function(func)
      .config(SeriesConfig.empty())
      .build();
  }

  static create(func: string, field?: string, parameter?: string | number) {
    const optionalParameter = parameter ? `,${parameter}` : '';
    const functionWithField = `${func}(${field ?? ''}${optionalParameter})`;

    return Series.forFunction(functionWithField);
  }

  toBuilder() {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    return new Builder(Immutable.Map(this._value));
  }
}

type BuilderState = Immutable.Map<string, any>;

class Builder {
  private readonly value: BuilderState;

  constructor(value: BuilderState = Immutable.Map()) {
    this.value = value;
  }

  function(newFunction: string) {
    return new Builder(this.value.set('function', newFunction));
  }

  config(newConfig: SeriesConfig) {
    return new Builder(this.value.set('config', newConfig));
  }

  parameter(newParameter: any) {
    return new Builder(this.value.set('parameter', newParameter));
  }

  build() {
    const { config } = this.value.toObject();
    const func = this.value.get('function');

    return new Series(func, config);
  }
}
