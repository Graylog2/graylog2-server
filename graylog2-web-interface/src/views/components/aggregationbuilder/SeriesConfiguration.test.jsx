import React from 'react';
import { mount } from 'wrappedEnzyme';

import Series from 'views/logic/aggregationbuilder/Series';
import SeriesConfig from 'views/logic/aggregationbuilder/SeriesConfig';
import SeriesConfiguration from './SeriesConfiguration';

describe('SeriesConfiguration', () => {
  const createNewSeries = (series = Series.forFunction('count()'), name) => {
    const newConfig = series.config.toBuilder().name(name).build();
    return series.toBuilder().config(newConfig).build();
  };

  it('renders the configuration dialog', () => {
    const wrapper = mount(<SeriesConfiguration series={createNewSeries()} onClose={() => {}} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders an input to change the series name', () => {
    const wrapper = mount(<SeriesConfiguration series={createNewSeries()} onClose={() => {}} />);
    expect(wrapper.find('input')).toHaveProp('value', 'count()');
  });

  it('uses the series\' effective value for the input', () => {
    const series = Series.forFunction('avg(took_ms)')
      .toBuilder()
      .config(SeriesConfig.empty().toBuilder().name('Average Request Time').build())
      .build();
    const wrapper = mount(<SeriesConfiguration series={series} onClose={() => {}} />);
    expect(wrapper.find('input')).toHaveProp('value', 'Average Request Time');
  });

  it('submit button calls onClose callback', () => {
    const onClose = jest.fn();
    const series = createNewSeries();
    const wrapper = mount(<SeriesConfiguration series={series} onClose={onClose} />);

    const submit = wrapper.find('button');

    submit.simulate('click');

    expect(onClose).toHaveBeenCalledTimes(1);
    expect(onClose).toHaveBeenCalledWith(series);
  });

  it('returns changed name upon submit', () => {
    const onClose = jest.fn();
    const series = createNewSeries();
    const wrapper = mount(<SeriesConfiguration series={series} onClose={onClose} />);
    const input = wrapper.find('input');
    const submit = wrapper.find('button');

    expect(input).toHaveProp('value', 'count()');

    input.simulate('change', { target: { value: 'Some other value' } });

    submit.simulate('click');

    const newSeries = createNewSeries(series, 'Some other value');
    expect(onClose).toHaveBeenCalledTimes(1);
    expect(onClose).toHaveBeenCalledWith(newSeries);
  });

  it('prevents entering a duplicate series name', () => {
    const onClose = jest.fn();
    const series = createNewSeries();
    const wrapper = mount(<SeriesConfiguration series={series} onClose={onClose} usedNames={['Already Exists']} />);
    const input = wrapper.find('input');

    expect(input).toHaveProp('value', 'count()');

    input.simulate('change', { target: { value: 'Already Exists' } });

    wrapper.update();
    const submit = wrapper.find('button');
    expect(submit).toBeDisabled();
    expect(wrapper).toIncludeText('Name must be unique');
  });

  it('returns original metric function name upon submit, when no name is defined', () => {
    const onClose = jest.fn();
    const series = createNewSeries(undefined, 'Some other value');
    const wrapper = mount(<SeriesConfiguration series={series} onClose={onClose} />);
    const input = wrapper.find('input');
    const submit = wrapper.find('button');

    expect(input).toHaveProp('value', 'Some other value');

    input.simulate('change', { target: { value: '' } });

    submit.simulate('click');

    const newSeries = createNewSeries(series, 'count()');
    expect(onClose).toHaveBeenCalledTimes(1);
    expect(onClose).toHaveBeenCalledWith(newSeries);
  });
});
