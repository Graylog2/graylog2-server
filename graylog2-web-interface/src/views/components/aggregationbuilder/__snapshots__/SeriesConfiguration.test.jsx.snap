// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SeriesConfiguration renders the configuration dialog 1`] = `
<SeriesConfiguration
  onClose={[Function]}
  series={
    Object {
      "config": Object {
        "name": undefined,
      },
      "function": "count()",
    }
  }
  usedNames={Array []}
>
  <span>
    <Component
      validationState={null}
    >
      <FormGroup__StyledFormGroup
        validationState={null}
      >
        <StyledComponent
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "SIZES": Array [
                "large",
                "small",
              ],
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "FormGroup__StyledFormGroup-sc-1wv4cm9-0",
                "isStatic": false,
                "lastClassName": "ekazNr",
                "rules": Array [
                  [Function],
                  ";",
                ],
              },
              "displayName": "FormGroup__StyledFormGroup",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "FormGroup__StyledFormGroup-sc-1wv4cm9-0",
              "target": [Function],
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
          validationState={null}
        >
          <FormGroup
            bsClass="form-group"
            className="FormGroup__StyledFormGroup-sc-1wv4cm9-0 ekazNr"
            validationState={null}
          >
            <div
              className="FormGroup__StyledFormGroup-sc-1wv4cm9-0 ekazNr form-group"
            >
              <ControlLabel>
                <StyledComponent
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "ControlLabel-sc-1edmum5-0",
                        "isStatic": false,
                        "lastClassName": "hCskeX",
                        "rules": Array [
                          [Function],
                        ],
                      },
                      "displayName": "ControlLabel",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "ControlLabel-sc-1edmum5-0",
                      "target": [Function],
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                >
                  <ControlLabel
                    bsClass="control-label"
                    className="ControlLabel-sc-1edmum5-0 hCskeX"
                    srOnly={false}
                  >
                    <label
                      className="ControlLabel-sc-1edmum5-0 hCskeX control-label"
                    >
                      Name
                    </label>
                  </ControlLabel>
                </StyledComponent>
              </ControlLabel>
              <FormControl
                onChange={[Function]}
                type="text"
                value="count()"
              >
                <StyledComponent
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "Feedback": [Function],
                      "SIZES": Array [
                        "small",
                        "large",
                      ],
                      "Static": [Function],
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "FormControl-sc-1amoaox-0",
                        "isStatic": false,
                        "lastClassName": "igyPGR",
                        "rules": Array [
                          [Function],
                        ],
                      },
                      "displayName": "FormControl",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "FormControl-sc-1amoaox-0",
                      "target": [Function],
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                  onChange={[Function]}
                  type="text"
                  value="count()"
                >
                  <FormControl
                    bsClass="form-control"
                    className="FormControl-sc-1amoaox-0 igyPGR"
                    componentClass="input"
                    onChange={[Function]}
                    type="text"
                    value="count()"
                  >
                    <input
                      className="FormControl-sc-1amoaox-0 igyPGR form-control"
                      onChange={[Function]}
                      type="text"
                      value="count()"
                    />
                  </FormControl>
                </StyledComponent>
              </FormControl>
              <HelpBlock>
                <StyledComponent
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "HelpBlock-i8is92-0",
                        "isStatic": false,
                        "lastClassName": "hCBNCb",
                        "rules": Array [
                          [Function],
                        ],
                      },
                      "displayName": "HelpBlock",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "HelpBlock-i8is92-0",
                      "target": [Function],
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                >
                  <HelpBlock
                    bsClass="help-block"
                    className="HelpBlock-i8is92-0 hCBNCb"
                  >
                    <span
                      className="HelpBlock-i8is92-0 hCBNCb help-block"
                    >
                      The name of the series as it appears in the chart.
                    </span>
                  </HelpBlock>
                </StyledComponent>
              </HelpBlock>
            </div>
          </FormGroup>
        </StyledComponent>
      </FormGroup__StyledFormGroup>
    </Component>
    <div
      className="pull-right"
      style={
        Object {
          "marginBottom": "10px",
        }
      }
    >
      <ForwardRef
        bsStyle="success"
        disabled={false}
        onClick={[Function]}
      >
        <Button__StyledButton
          bsStyle="success"
          disabled={false}
          onClick={[Function]}
        >
          <StyledComponent
            bsStyle="success"
            disabled={false}
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "SIZES": Array [
                  "large",
                  "small",
                  "xsmall",
                ],
                "STYLES": Array [
                  "success",
                  "warning",
                  "danger",
                  "info",
                  "default",
                  "primary",
                  "link",
                ],
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "Button__StyledButton-c9cbmb-0",
                  "isStatic": false,
                  "lastClassName": "dRMLmK",
                  "rules": Array [
                    [Function],
                  ],
                },
                "displayName": "Button__StyledButton",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "Button__StyledButton-c9cbmb-0",
                "target": [Function],
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            onClick={[Function]}
          >
            <Button
              active={false}
              block={false}
              bsClass="btn"
              bsStyle="success"
              className="Button__StyledButton-c9cbmb-0 dRMLmK"
              disabled={false}
              onClick={[Function]}
            >
              <button
                className="Button__StyledButton-c9cbmb-0 dRMLmK btn btn-success"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                Done
              </button>
            </Button>
          </StyledComponent>
        </Button__StyledButton>
      </ForwardRef>
    </div>
  </span>
</SeriesConfiguration>
`;
